### PR TITLE
Use cross-module R2R generics from main module with version bubble

### PR DESCRIPTION
**Pull request by fadimounir**
Porting the changes from the single-exe branch, and fixing a bug with the native assembly manifest where we had missing entries (any module directly referenced by the current module).

cc @dotnet/crossgen-contrib

### CODE REVIEW

**Comment by fadimounir** [comment]
It would have been nice to have some test hooks similar to the ngen zaprequire family to test R2R version bubble validation code (ex: verify certain R2R images get used/rejected, and verify certain R2R code gets used/rejected). It would have been possible to add a test case for these changes...

I'll add this on my list of TODOs.

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ManifestMetadataTableNode.cs [hunk 1]
```diff
@@ -95,13 +101,28 @@ public int ModuleToIndex(EcmaModule module)
...
 
+                // If we're going to add a module to the manifest, it has to be part of the version bubble, otherwise
+                // the verification logic would be broken at runtime.
+                Debug.Assert(_nodeFactory.CompilationModuleGroup.VersionsWithModule(module));
+
                 assemblyRefIndex = _nextModuleId++;
                 _manifestAssemblies.Add(assemblyName);
                 _assemblyRefToModuleIdMap.Add(assemblyName.Name, assemblyRefIndex);
             }
+            else if (_nodeFactory.CompilationModuleGroup.VersionsWithModule(module))
```
Comment (trylek) [inline src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ManifestMetadataTableNode.cs]: Please consider unifying both conditional blocks - the unification should be in the above block too.

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs [hunk 2]
```diff
@@ -29,6 +31,11 @@ public override bool ContainsType(TypeDesc type)
             return type == _method.OwningType;
         }
 
+        public override bool VersionsWithModule(ModuleDesc module)
+        {
+            return ((EcmaMethod)_method).Module == module;
```
Comment (trylek) [inline src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Should we use GetTypicalMethodDefinition?

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ManifestMetadataTableNode.cs (hunk 1 above)
Comment (fadimounir) [reply src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ManifestMetadataTableNode.cs]: Thanks for the suggestion Tomas. Fixed

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs (hunk 2 above)
Comment (fadimounir) [reply src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: I think _method and its typical definition will have the same module

Comment (trylek) [reply src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: It will but a generic method won't cast to an EcmaMethod

Comment (fadimounir) [reply src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/SingleMethodCompilationModuleGroup.cs]: Ah, good catch. I'll fix that.

**Pull request merged**